File: src/test/java/j/math/cat/FunctorTest.java
package j.math.cat;

import static j.math.cat.BasePair.Pair;
import static j.math.cat.Category.Category;
import static j.math.cat.Sets.Set;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import junit.framework.TestCase;

/**
 * Unittest for Functor class.
 * 
 * @author Vlad Patryshev
 * All source code is stored on <a href="http://code.google.com/p/categories/">http://code.google.com/p/categories/</a>
 *
 */
public class FunctorTest extends TestCase {
  private final Category<String, String> categoryTwoToOne =
      Category("(([0,1,2], {a: 0 -> 2, b: 1 -> 2}))");

  private final Category<String, String> categorySquareWithTwoTopLeftCorners =
      Category("(([a0,a1,b,c,d], {a0a1: a0 -> a1, a0b: a0 -> b, a0c: a0 -> c, a1b: a1 -> b, a1c: a1 -> c, bd: b -> d, cd: c -> d, a0d: a0 -> d, a1d: a1 -> d}), {bd o a0b = a0d, cd o a0c = a0d, bd o a1b = a1d, cd o a1c = a1d, a1b o a0a1 = a0b, a1c o a0a1 = a0c, a1d o a0a1 = a0d})");

  Functor<String, String, String, String> functorFrom2to1toDoubleSquare =
      Functor.Functor(categoryTwoToOne,
              categorySquareWithTwoTopLeftCorners,
              Base.Map(new String[]{"0", "1", "2"},
                      new String[]{"b", "c", "d"}),
              Base.Map(new String[]{"0", "1", "2", "a", "b"},
                      new String[]{"b", "c", "d", "bd", "cd"}));

  Category<String, String> category1to2 =
      Category("(([0,1,2], {a: 0 -> 1, b: 0 -> 2}))");

  Category<String, String> categorySquareWithTwoRightCorners =
      Category(
          "(([a,b,c,d0,d1], " +
              "{ab: a -> b, ac: a -> c, ad0: a -> d0, bd0: b -> d0, cd0: c -> d0, ad1: a -> d1, bd1: b -> d1, cd1: c -> d1, d0d1: d0 -> d1}), " +
              "{bd0 o ab = ad0, cd0 o ac = ad0, bd1 o ab = ad1, cd1 o ac = ad1, d0d1 o ad0 = ad1, , d0d1 o bd0 = bd1, d0d1 o cd0 = cd1})");

  Functor<String, String, String, String> functorFrom1to2toDoubleSquare =
      Functor.Functor(category1to2, categorySquareWithTwoRightCorners,
              Base.Map(new String[]{"0", "1", "2"},
                      new String[]{"a", "b", "c"}),
              Base.Map(new String[]{"0", "1", "2", "a", "b"},
                      new String[]{"a", "b", "c", "ab", "ac"}));

  public void testFunctor_missingObjectMapppings() {
    try {
      Functor.Functor(Categories._4_, Categories._4_, Base.Map(new Integer[]{0}, new Integer[]{1}),
              new HashMap<Pair<Integer, Integer>, Pair<Integer, Integer>>()
      );
      fail("This constructor should have failed");
    } catch (Throwable e) {
      // as expected
    }
  }

  public void testFunctor_missingArrowMapppings() {
    try {
      Functor.Functor(Categories._4_, Categories._4_,
              Base.Map(
                      new Integer[]{0, 1, 2, 3},
                      new Integer[]{1, 2, 3, 3}),
              new HashMap<Pair<Integer, Integer>, Pair<Integer, Integer>>()
      );
      fail("This constructor should have failed");
    } catch (Throwable e) {
      // as expected
    }
  }

  @SuppressWarnings("serial")
  public void testFunctor_unitBroken() {
    try {
      Functor.Functor(Categories._4_, Categories._4_,
              new HashMap<Integer, Integer>() {
                  {
                      put(0, 1);
                      put(1, 2);
                      put(2, 3);
                      put(3, 3);
                  }
              },
              new HashMap<Pair<Integer, Integer>, Pair<Integer, Integer>>() {
                  {
                      put(BasePair.Pair(0, 0), BasePair.Pair(1, 1));
                      put(BasePair.Pair(0, 1), BasePair.Pair(1, 2));
                      put(BasePair.Pair(0, 2), BasePair.Pair(1, 3));
                      put(BasePair.Pair(0, 3), BasePair.Pair(1, 3));
                      put(BasePair.Pair(1, 1), BasePair.Pair(2, 3));
                      put(BasePair.Pair(1, 2), BasePair.Pair(2, 3));
                      put(BasePair.Pair(1, 3), BasePair.Pair(2, 3));
                      put(BasePair.Pair(2, 2), BasePair.Pair(3, 3));
                      put(BasePair.Pair(2, 3), BasePair.Pair(3, 3));
                      put(BasePair.Pair(3, 3), BasePair.Pair(3, 3));
                  }
              }
      );
      fail("This constructor should have failed");
    } catch (Throwable e) {
      // as expected
    }
  }

  @SuppressWarnings("serial")
  public void testFunctor_plain() {
    try {
      Functor.Functor(Categories._4_, Categories._4_,
              new HashMap<Integer, Integer>() {
                  {
                      put(0, 1);
                      put(1, 2);
                      put(2, 3);
                      put(3, 3);
                  }
              },
              new HashMap<Pair<Integer, Integer>, Pair<Integer, Integer>>() {
                  {
                      put(BasePair.Pair(0, 0), BasePair.Pair(1, 1));
                      put(BasePair.Pair(0, 1), BasePair.Pair(1, 2));
                      put(BasePair.Pair(0, 2), BasePair.Pair(1, 3));
                      put(BasePair.Pair(0, 3), BasePair.Pair(1, 3));
                      put(BasePair.Pair(1, 1), BasePair.Pair(2, 2));
                      put(BasePair.Pair(1, 2), BasePair.Pair(2, 3));
                      put(BasePair.Pair(1, 3), BasePair.Pair(2, 3));
                      put(BasePair.Pair(2, 2), BasePair.Pair(3, 3));
                      put(BasePair.Pair(2, 3), BasePair.Pair(3, 3));
                      put(BasePair.Pair(3, 3), BasePair.Pair(3, 3));
                  }
              }
      );
      fail("This constructor should have failed");
    } catch (Throwable e) {
      // as expected
    }
  }

  @SuppressWarnings("serial")
  public void testLimit_cartesianProductActually() {
    Category<Integer, Integer> from = Categories.discreteCategory(Sets.numbers(2));
    Category<String, String> to = Categories.SQUARE;
    Map<Integer, String> map =
        new HashMap<Integer, String>() {
          {
            put(0, "b");
            put(1, "c");
          }
        };

    Functor<Integer, Integer, String, String> f = Functor.Functor(from, to, map, map);
    Functor.Cone limit = f.limit();
    assertNotNull(limit);
    TestCase.assertEquals("ab", limit.arrowTo(0));
    TestCase.assertEquals("ac", limit.arrowTo(1));
  }

  @SuppressWarnings({ "serial" })
  public void testAllCones() {
    Set<Functor<String,String,String,String>.Cone> allCones =
        functorFrom2to1toDoubleSquare.allCones();
    Functor.Cone c1 =
        functorFrom2to1toDoubleSquare.new Cone("a0",
            new HashMap<String, String>() {
              {
                put("0", "a0b");
                put("1", "a0c");
                put("2", "a0d");
              }
            });
    Functor.Cone c2 =
        functorFrom2to1toDoubleSquare.new Cone("a1",
            new HashMap<String, String>() {
              {
                put("0", "a1b");
                put("1", "a1c");
                put("2", "a1d");
              }
            }
        );
    Set<Functor.Cone> expectedCones = Sets.Set(c1, c2);
    assertEquals(expectedCones.size(), allCones.size());
    assertTrue("c1 should belong", allCones.contains(c1));
    assertTrue("c2 should belong", allCones.contains(c1));
    for (Functor.Cone c: allCones) {
      assertTrue("this cone: " + c + " is unexpected", expectedCones.contains(c));
    }
    assertEquals(expectedCones, allCones);
  }

  public void testLimit_twoCandidates() {
    Functor.Cone limit = functorFrom2to1toDoubleSquare.limit();
    assertNotNull(limit);
    TestCase.assertEquals("a1", limit.apex());
    TestCase.assertEquals("a1b", limit.arrowTo("0"));
    TestCase.assertEquals("a1c", limit.arrowTo("1"));
  }

  @SuppressWarnings({ "serial"})
  public void testCoconesTo() {
    Set<Functor<String,String,String,String>.Cocone> actual =
        functorFrom1to2toDoubleSquare.coconesTo("d0");
    Functor.Cocone expected =
        functorFrom1to2toDoubleSquare.new Cocone("d0",
            new HashMap<String, String>() {
              {
                put("0", "ad0");
                put("1", "bd0");
                put("2", "cd0");
              }
            });
    TestCase.assertEquals(Sets.Set(expected), actual);
  }

  @SuppressWarnings({ "serial" })
  public void testAllCocones() {
    Set<Functor<String,String,String,String>.Cocone> allCocones =
        functorFrom1to2toDoubleSquare.allCocones();
    Functor.Cocone c1 =
        functorFrom1to2toDoubleSquare.new Cocone("d0",
            new HashMap<String, String>() {
              {
                put("0", "ad0");
                put("1", "bd0");
                put("2", "cd0");
              }
            });
    Functor.Cocone c2 =
        functorFrom1to2toDoubleSquare.new Cocone("d1",
            new HashMap<String, String>() {
              {
                put("0", "ad1");
                put("1", "bd1");
                put("2", "cd1");
              }
            }
        );
    Set<Functor.Cocone> expectedCocones = Sets.Set(c1, c2);
    for (Functor.Cocone cocone : allCocones) {
      assertTrue(cocone.equals(c1) || cocone.equals(c2));
    }
    assertEquals(expectedCocones, allCocones);
  }

  public void testColimit_twoCandidates() {
    Pair<String, Map<String, String>> colimit = functorFrom1to2toDoubleSquare.colimit();
    assertNotNull(colimit);
    assertEquals("d0", colimit.x());
    assertEquals("ad0", colimit.y().get("0"));
    assertEquals("bd0", colimit.y().get("1"));
    assertEquals("cd0", colimit.y().get("2"));
  }
}

File: scala3/src/test/scala/math/cat/FunctorTest.scala
package math.cat

import math.Test
import math.cat.Categories._
import math.cat.SetCategory.Setf
import math.sets.Sets.set
import org.specs2.matcher.MatchResult
import scalakittens.Good

import scala.language.postfixOps
import SetFunction.fun

class FunctorTest extends Test:
  type SUT = Functor
  
  lazy val categorySquareWithTwoTopLeftCorners: Cat =
    category"({a0,a1,b,c,d}, {a0a1: a0 -> a1, a0b: a0 -> b, a0c: a0 -> c, a1b: a1 -> b, a1c: a1 -> c, bd: b -> d, cd: c -> d, a0d: a0 -> d, a1d: a1 -> d}, {bd ∘ a0b = a0d, cd ∘ a0c = a0d, bd ∘ a1b = a1d, cd ∘ a1c = a1d, a1b ∘ a0a1 = a0b, a1c ∘ a0a1 = a0c, a1d ∘ a0a1 = a0d})"

  lazy val functorFromPullbackToDoubleSquare: SUT = {
    Functor(
      "From2to1toDoubleSquare",
      Pullback, categorySquareWithTwoTopLeftCorners)(
      Map("a" -> "b", "b" -> "c", "c" -> "d"),
      Map("a" -> "b", "b" -> "c", "c" -> "d", "ac" -> "bd", "bc" -> "cd")
    ) getOrElse (throw new InstantiationException("Something wrong with `From2to1toDoubleSquare`"))
  }

  lazy val categorySquareWithTwoRightCorners =
    category"""({a,b,c, d0, d1}
      ,{ab: a -> b, ac: a -> c, ad0: a -> d0, bd0: b -> d0, cd0: c -> d0, ad1: a -> d1, bd1: b -> d1, cd1: c -> d1, d0d1: d0 -> d1}
      ,{bd0 ∘ ab = ad0, cd0 ∘ ac = ad0, bd1 ∘ ab = ad1, cd1 ∘ ac = ad1, d0d1 ∘ ad0 = ad1, d0d1 ∘ bd0 = bd1,d0d1 ∘ bd0 = bd1, d0d1 ∘ cd0 = cd1}
      )"""

  lazy val functorFrom1to2toDoubleSquare: SUT = {

    Functor("From1to2toDoubleSquare",
      Pushout, categorySquareWithTwoRightCorners)(
      Map("a" -> "a", "b" -> "b", "c" -> "c"),
      Map("a" -> "a", "b" -> "b", "c" -> "c", "ab" -> "ab", "ac" -> "ac")
    ).iHope
  }
  
  "Constructor" should {

    "report missing object mappings" in:
      expectError("Object mapping fails for 1",
      Functor("failing test",
        _4_, _4_)(
        Map("0" -> "1"),
        Map.empty[String, String]))

    "report incorrect object mappings" in:
      expectError("Object mapping fails for 1",
        Functor("failing test",
          _2_, _2_)(
          Map("0" -> "1", "1" -> "3"),
          Map.empty[String, String]))

    "report missing arrows mappings" in:
      expectError("Missing arrow mappings for 0.1, 1.2, 2.3, 0.3, 1.3, 0.2",
        Functor("failing test",
        _4_, _4_)(
        Map("0" -> "1", "1" -> "2", "2" -> "3", "3" -> "3"),
          Map.empty[String, String]))

    "report inconsistent arrow mappings" in:
      val objectMapping: _4_.Node => _4_.Node =
        Map[_4_.Node, _4_.Node]("0" -> "1", "1" -> "2", "2" -> "1", "3" -> "3")
      val arrowMapping: _4_.Arrow => _4_.Arrow = Map[_4_.Arrow, _4_.Arrow](
        "0.0" -> "1.1",
        "0.1" -> "1.2",
        "0.2" -> "1.3",
        "0.3" -> "1.3",
        "1.1" -> "2.3",
        "1.2" -> "2.3",
        "1.3" -> "2.3",
        "2.2" -> "3.3",
        "2.3" -> "3.3",
        "3.3" -> "3.3"
      )
      expectError("Inconsistent mapping for d1(0.2)",
        Functor("id mapping broken", _4_, _4_)(objectMapping, arrowMapping))

    "report a failure" in:
      val objectMapping: _4_.Node => _4_.Node =
        Map[_4_.Node, _4_.Node]("0" -> "1", "1" -> "2", "2" -> "3", "3" -> "4")
      val arrowMapping = Map(
        "0.0" -> "1.1",
        "0.1" -> "1.2",
        "0.2" -> "1.3",
        "0.3" -> "1.3",
        "1.1" -> "2.2",
        "1.2" -> "2.3",
        "1.3" -> "2.3",
        "2.2" -> "3.3",
        "2.3" -> "3.3",
        "3.3" -> "3.3"
      )

      expectError("Object mapping fails for 3",
        Functor("something wrong here", _4_, _4_)(objectMapping, arrowMapping))
  }

  "functor" should {
    "produce cartesian product" in {
      val from = Category.discrete(Set(0, 1))
      val to = Square

      val mapA: from.Arrow => to.Arrow =
        Map(0 -> "b", 1 -> "c")

      type toO = to.Obj
      val mapO: from.Obj => to.Obj =
        Map(0 -> "b", 1 -> "c")

      val fOpt = Functor("sample product", from, to)(mapO, mapA)
      checkOption[Functor](fOpt,
        (f: Functor) =>
          val limitOpt = f.limit

          limitOpt match
            case Good(limit) =>
              limit.arrowTo(0) == "ab" && limit.arrowTo(1) == "ac"
            case bad => failure(s"Could not build a limit of $f: $bad")
      )
    }
    
    "cones from" in {
      val sut = functorFromPullbackToDoubleSquare
      val obj0 = sut.d0.obj
      val obj1 = sut.d1.obj
      val arr1 = sut.d1.asArrow
      val actual: Set[sut.Cone] = sut.conesFrom(obj1("a0"))
      
      val expected: sut.Cone = sut.Cone(obj1("a0"), Map(
        obj0("a") -> arr1("a0b"),
        obj0("b") -> arr1("a0c"),
        obj0("c") -> arr1("a0d")
      ))
        
      actual === Set(expected)
    }
    
    "build all cones" in {
      val sut = functorFromPullbackToDoubleSquare
      val obj0 = sut.d0.obj
      val obj1 = sut.d1.obj
      val arr1 = sut.d1.asArrow
      val allCones = sut.allCones
      val c1 = sut.Cone(obj1("a0"), Map(
        obj0("a") -> arr1("a0b"),
        obj0("b") -> arr1("a0c"),
        obj0("c") -> arr1("a0d")
      ))
      
      val c2 = sut.Cone(obj1("a1"), Map(
        obj0("a") -> arr1("a1b"),
        obj0("b") -> arr1("a1c"),
        obj0("c") -> arr1("a1d")
      ))
      
      allCones === Set(c1, c2)
    }
    
    "limit with two candidates" in {
      val sut = functorFromPullbackToDoubleSquare
      sut.limit match
        case Good(limit) =>
          limit.vertex === "a1"
          limit.arrowTo("a") === "a1b"
          limit.arrowTo("b") === "a1c"
        case oops => failure("no limit?")

      ok
    }
    
    "cocones to" in {
      val sut = functorFrom1to2toDoubleSquare
      val obj0 = sut.d0.obj
      val obj1 = sut.d1.obj
      val arr1 = sut.d1.asArrow

      val actual = sut.coconesTo(obj1("d0"))
      val expected = sut.Cocone(obj1("d0"), Map(
        obj0("a") -> arr1("ad0"),
        obj0("b") -> arr1("bd0"),
        obj0("c") -> arr1("cd0")
      ))
      actual === Set(expected)
    }
    
    "all cocones" in {
      val sut = functorFrom1to2toDoubleSquare
      val obj0 = sut.d0.obj
      val obj1 = sut.d1.obj
      val arr1 = sut.d1.asArrow
      val allCocones = sut.allCocones
      val expected1 = sut.Cocone(obj1("d0"), Map(
        obj0("a") -> arr1("ad0"),
        obj0("b") -> arr1("bd0"),
        obj0("c") -> arr1("cd0")
      ))
      val expected2 = sut.Cocone(obj1("d1"), Map(
        obj0("a") -> arr1("ad1"),
        obj0("b") -> arr1("bd1"),
        obj0("c") -> arr1("cd1")
      ))
      allCocones === Set(expected1, expected2)
    }
    
    "colimit with two candidates" in {
      val sut = functorFrom1to2toDoubleSquare
      val obj0 = sut.d0.obj
      val obj1 = sut.d1.obj
      val arr1 = sut.d1.asArrow

      sut.colimit match
        case Good(colimit) =>
          colimit.vertex === "d1"
          colimit.arrowFrom(obj0("a")) === "ad1"
          colimit.arrowFrom(obj0("b")) === "bd1"
          colimit.arrowFrom(obj0("c")) === "cd1"
        case oops => failure("Could not build a colimit for " + 
                     functorFrom1to2toDoubleSquare.tag)

      ok
    }

    "validate with Set as domain" in {
      val a: set = Set(1, 2, 3)
      val b: set = Set(2, 3, 4)
      val c: set = Set(0, 1)
      val ac = fun(a,c)("f", _.toInt % 2)
      val bc = fun(b,c)("g", x => (x.toInt + 1) % 2)
      val sutOpt = Functor(
        "pullback", Pullback, SetCategory.Setf)(
        Map("a" -> a, "b" -> b, "c" -> c),
        Map("ac" -> ac, "bc" -> bc)
      )
      
      checkOption[Functor](sutOpt,
        sut => {
          sut.d0 === Pullback
          sut.d1 === Setf
          sut.objectsMapping("a") === a
          sut.objectsMapping("b") === b
          sut.objectsMapping("c") === c
          sut.arrowsMapping("ac") === ac
        }
      )
    }
  }
